_posts/2023-01-23-there-is-no-cloud.md

```diff
@@ -0,0 +1,294 @@
+provider to decrypt hosted data. Note that you keep your keys as well
+as a client computer (e.g., your laptop) strictly out of the cloud;
+otherwise your effort is immediately compromised.
+
+Is client-side encryption mainstream? There are lots of open-source
+cryptographic libraries you can use to encrypt your data before
+submitting to a cloud, so you are all set. But can you 'transparently'
+integrate client-side encryption with, say, cloud databases?
+Transparently means you spend little or no effort on en/decrypting,
+just do your usual database CRUD operations. Here we have quite a
+limited choice.
+
+
+## AWS
+
+AWS offers DynamoDB Encryption Client in Java and Python only. You can
+use these SDKs with AWS KMS (no real privacy) or provide your own
+(full privacy) cryptographic material for encryption (you keep away
+from the cloud). Unfortunately, at the time of writing, relational
+databases in AWS lack client-side encryption support.
+
+
+## Azure
+
+How about Azure CosmosDB? They created the so-called Always Encrypted
+client-side option for CosmosDB, in .NET and Java. Unfortunately, at
+the time of writing, the only option to keep your customer-managed
+keys CMK for client-side encryption (which should be your most
+valuable private property) suggested by Microsoft is the Azure
+KeyVault, which undermines the whole privacy idea (goats to wolves) 
+straight away.
+
+
+## Primary Keys Are NOT Encrypted
+
+It is important to note that both AWS DynamoDB and Azure CosmosDB
+require that you always leave the primary keys (partition/hash and
+range/sort keys) **unencrypted**. The motivation is that NoSQL databases
+need primary keys for proper and efficient storing and fetching data
+(think of the necessity to repartition your DB when encryption keys
+are rotated and encrypted primary keys change respectively). This
+unencrypted primary keys requirement is necessary but a bit
+disappointing, since it exposes part of your important data. If you
+ever worked with NoSQL databases, you realize that normally primary
+keys are composed of many other attributes to make searches efficient,
+which exposes your data even more. For example, you may use name,
+surname, and DOB as primary keys (some exposure) but use Bitcoin
+credentials (extremely sensitive) for client-side encryption. 
+Another example is you use a personal number as a(n unencrypted) 
+primary key, but keep all other patient information fully client-side
+encrypted. Full client-side encryption is of course possible if you 
+store and operate your DB completely outside the cloud, then encrypt 
+and upload it to the cloud. This is possible with the cloud blob 
+storage in any cloud.
+
+
+## GCP
+
+In contrast to AWS and Azure, GCP offers client-side encryption for
+Cloud SQL.
+
+
+## Conclusions
+
+Client-side encryption options are available, but not abundant. Let's 
+expect more will emerge, since motivation is pretty obvious.
+
```

Comment:
what about the thing we most often use in projects, regular relational SQL databases? any solutions for  those? how about text search engines, those probably won't work if your data is encrypted.

Reply:
we meaning Solita and our typical projects... but consider adding something about SQL in the blog post.

Reply:
There is a short mention about client-side encryption in GCP Cloud SQL

Reply:
If there are questions about SQL, I'll need to work out the details about the GCP Cloud SQL with client-side encryption. But this will take another bog post. Most likely, I will provide full code